docs/DevelopersGuide.adoc

```diff
@@ -47,333 +49,1007 @@ the Datomic pull syntax):
+;; Define one or more resolvers
+(defresolver `some-symbol ...)
+(defresolver `some-other-symbol ...)
+...
 
-We will cover examples of all of these in this chapter. More recent versions of pathom include `Connect`.
-It's a new approach to define a mechanism for *Computed attributes* that we will talk about later.
+;; Create a parser that uses the resolvers:
+(def parser
+  (p/parser {::p/plugins
+             [(p/env-plugin
+                {::p/reader             [p/map-reader pc/all-readers]
+                 ::pc/resolver-dispatch resolver-fn
+                 ::pc/indexes           @indexes})]}))
+----
 
-IMPORTANT: You should take great care to use unique names for your graph query attributes. The contextual behavior can lead to confusion very quickly, and it is made worse if you're not sure which part of your query is even running. Also, software evolution will almost certainly lead you to use your queries with new APIs, and you will want to avoid naming collisions. Use qualified (ideally namespaced) keywords.
+== Resolvers
 
-== A Parser with Entity Lookups and Computed Attributes
+In `Connect` you implement the graph by creating `resolvers`, those resolvers are functions that expose some data on the graph.
 
-Next, let's show an example of using entity and computed attributes together.
+A resolver has a few basic elements:
 
-First, let's talk about the parsing environment a little more. Typically the parsing environment will include potentially dynamic things (e.g. a database connection) and some consistent things (e.g. the reader).
+. Inputs – A set of attributes that must be in the current parsing context for the resolver to be able to work.  Inputs is
+optional, and missing inputs means that the resolver is always capable of working independent of the current parsing context.
+. Outputs - A query-like notation that gives the "pattern" of the part of the query the resolver is able to resolve. This is typically a list of attributes/joins, where joins typically include a simple subquery.
+. A lambda - A `(fn [env input-data] tree-of-promised-output)` that takes the inputs and turns them into a tree that satisfies the "output query".
 
-In the prior example we explicitly added the reader to the environment every time we called the parser. In cases where there are specific things that you'd always like included in the environment we can instead use the plugin system to pre-set them.
+NOTE: Resolvers *must* always return a map with at least one key. We'll cover how to indicate "not found" in a bit.
 
-In our prior example we had:
+So you might define a resolver like this:
 
 ```
-(def parser (p/parser {}))
+(defresolver `person-resolver
+  {::pc/input #{:person/id}
+   ::pc/output [:person/name :person/age]}
+  (fn [{:keys [database] :as env} {:keys [person/id]}]
+    (let [person (my-database/get-person database id)]
+      {:person/age (:age person)
+       :person/name (str (:first-name person) " " (:last-name person))})))
```

Comment:
This is another example we can improve on, would be better to get on peoples minds that any computation is better suited in a separated resolver. This makes for also a more flexible system, since you always can access the raw and write something on top of it. So to illustrate this example we could use names like `:fname` and `:lname` on the source and convert to `:person/first-name` and `:person/last-name` respectively. Then another resolve could work on those attributes to create a `:person/full-name`. This will show people that the source of that data doens't matter, but you can always augment based on the attribute names.

Reply:
So, your point is one resolver should be making the raw props (possibly renamed) available, and a different resolver could be written for `:person/full-name` which does the computation, but only if it was actually asked for.